web_app/contract_tools/ekubo_research.py

```diff
@@ -0,0 +1,203 @@
+"""
+Ekubo Flash Loan Research and Implementation Guide
+================================================
+
+This research document outlines how to implement flash loans on Ekubo AMM platform,
+based on their documentation and implementation examples.
+
+Key Concepts:
+------------
+1. Flash Accounting: Ekubo's internal token balance accounting system.
+2. Free Flash Loans: No fees for borrowing within same transaction.
+3. Atomic Execution: All operations must complete in single transaction.
+
+Implementation Flow:
+------------------
+
+API Endpoints(https://mainnet-api.ekubo.org/openapi.json) and Data Structures
+"""
+
+from dataclasses import dataclass
+from typing import List, Dict
+from decimal import Decimal
+import httpx
+
+@dataclass
+class PoolKey:
+    """Ekubo pool identification structure"""
+    token0: str
+    token1: str
+    fee: int
+    tick_spacing: int
+    extension: str
+
+@dataclass
+class SwapRoute:
+    """Structure for defining swap path"""
+    pool_key: PoolKey
+    sqrt_ratio_limit: int
+    skip_ahead: int
+
+@dataclass
+class TokenAmount:
+    """Token amount with direction"""
+    token: str
+    amount: int
+    is_positive: bool
+
+class EkuboFlashLoan:
+    """
+    Main class for handling Ekubo Flash Loan operations
+    
+    Key Features:
+    - No upfront fees required
+    - Must repay within same transaction
+    - Uses flash accounting for internal balance management
+    """
+
+    def __init__(self, rpc_url: str, core_contract: str):
+        self.rpc_url = rpc_url
+        self.core_contract = core_contract
+
+    async def get_profitable_route(self, input_token: str, amount: Decimal) -> List[SwapRoute]:
```

Comment:
Why not expect an amount already scaled with decimals, as it's required for both `get_profitable_route` and `prepare_flash_loan_tx` methods? With this approach, we'll need decimals only in a single place.